www/content/workstation/upgrade_labs.md

```diff
@@ -0,0 +1,330 @@
+
+This is not an ideal practice.  You will likely be making changes to the cookbooks in the steps ahead. It is important that you be able to track those changes and test your changes in a continuous integration pipeline, which is beyond the scope of this document. If you make changes in a copy of the cookbook without version control information, it will be difficult to reconcile those changes in the future. If you find yourself in this situation, it is likely worth the effort to try to track down the version-controlled source.
+
+```
+Changes made to the following cookbooks in ./node-MYNODE-repo/cookbooks
+cannot be saved upstream, though they can still be uploaded
+to a Chef Server:
+
+  - logrotate (v1.9.2)
+  - windows (v1.44.1)
+  - chef_handler (v1.4.0)
+
+You're ready to begin!
+
+Start with 'kitchen converge'.  As you identify issues, you
+can modify cookbooks in their original checkout locations or
+in the repository's cookbooks directory and they will be picked
+up on subsequent runs of 'kitchen converge'.
+```
+
+## Converge Locally
+
+{{< note >}}
+Be cautious when running your cookbooks locally. It is possible to accidentally impact production infrastructure based on settings in your cookbooks and attributes. Read over your cookbooks and attributes, especially attributes being set in roles and environments.  If needed, override attributes to be appropriate for local testing in your `kitchen.yml`.
+{{< /note >}}
+
+### Increment the Chef Infra Client Version
+
+In the `kitchen.yml` file, change the `product_version` line to `16`:
+```yaml
+provisioner:
+  name: chef_zero
+  product_name: chef
+  product_version: 16  # Change this line
+```
+
+If needed, you can "step forward" by first going from 12 to 13, correcting issues, then 13 to 14, etc.
+
+### Attempt a Converge and Check for Errors
+
+Run, in the `node-MY_NODE-repo` directory:
+```
+$ chef exec kitchen converge
+```
+Watch for Chef Infra errors. If any occur, fix them.  Also consider running cookstyle (see below). Repeat as needed.
+
+### Run CookStyle
+To check for version upgrade issues, run:
+```
+$ chef exec cookstyle cookbooks/some-cookbook
+```
+Repeat this process for each cookbook that the node consumes.
+
+## Correct any Cookbook Style Issues
+### Using auto-correct
+To auto-correct cookbook issues, run:
+```
+$ chef exec cookstyle -a cookbooks/some-cookbook
+```
+Other issues may require manual intervention and editing.
+Repeat this process for each cookbook that the node consumes.
+
+### Using the VSCode Plugin
+While any editor can be used, the Chef Infra extension for Visual Studio Code features several code generators and helpful features, such as running cookstyle when recipes are saved.
+
+### Check for Data Bags
+If data bags are used, you will need a `data_bags` directory in your repo.
+Pull down the data_bags by running:
+```
+$ chef exec knife download data_bags
```

Comment:
Needs --profile old-server flag

Reply:
^^ Didn't resolve issue.

```
 jobrien@joshs-mbp  ~/.chef  chef exec knife download data_bags
/opt/chef-workstation/embedded/lib/ruby/gems/2.7.0/gems/chef-15.10.12/lib/chef/resource.rb:1384: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call
/opt/chef-workstation/embedded/lib/ruby/gems/2.7.0/gems/chef-15.10.12/lib/chef/node_map.rb:58: warning: The called method `set' is defined here
Created data_bags
WARN: Failed to read the private key /etc/chef/client.pem: #<Errno::EACCES: Permission denied @ rb_sysopen - /etc/chef/client.pem>
ERROR: Your private key could not be loaded from /etc/chef/client.pem
Check your configuration file and ensure that your private key is readable
 ✘ jobrien@joshs-mbp  ~/.chef  chef exec knife download data_bags --profile old-server
/opt/chef-workstation/embedded/lib/ruby/gems/2.7.0/gems/chef-15.10.12/lib/chef/resource.rb:1384: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call
/opt/chef-workstation/embedded/lib/ruby/gems/2.7.0/gems/chef-15.10.12/lib/chef/node_map.rb:58: warning: The called method `set' is defined here
WARN: Failed to read the private key /etc/chef/client.pem: #<Errno::EACCES: Permission denied @ rb_sysopen - /etc/chef/client.pem>
ERROR: Your private key could not be loaded from /etc/chef/client.pem
Check your configuration file and ensure that your private key is readable
```

Reply:
Turned out the upload/download commands do not support keys embedded in credentials. I'm now shipping the extracted keys in the tarball and will update the onboarding doc.